helix-term/src/commands.rs

```diff
@@ -2325,11 +2320,49 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
 }
 
 #[inline]
-fn delete_selection_insert_mode(doc: &mut Document, view: &mut View, selection: &Selection) {
-    let transaction = Transaction::change_by_selection(doc.text(), selection, |range| {
-        (range.from(), range.to(), None)
-    });
+fn delete_by_selection_insert_mode(
+    cx: &mut Context,
+    mut f: impl FnMut(RopeSlice, &Range) -> Deletion,
+    direction: Direction,
+) {
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let mut selection = SmallVec::new();
+    let mut insert_newline = false;
+    let text_len = text.len_chars();
+    let mut transaction =
+        Transaction::delete_by_selection(doc.text(), doc.selection(view.id), |range| {
+            let (start, end) = f(text, range);
+            if direction == Direction::Forward {
+                let mut range = *range;
+                if range.head > range.anchor {
+                    insert_newline |= end == text_len;
+                    // move the cursor to the right so that the selection
+                    // doesn't shrink when deleting forward (so the text appears to
+                    // move to  left)
+                    // += 1 is enough here as the range is normalized to grapheme boundaries
+                    // later anyway
+                    range.head += 1;
+                }
+                selection.push(range);
+            }
+            (start, end)
+        });
+
+    // in case we delete the last character and the cursor would be moved to the EOF char
+    // insert a newline, just like when entering append mode
+    if insert_newline {
+        transaction = transaction.insert_at_eof(doc.line_ending.as_str().into());
+    }
+
+    if direction == Direction::Forward {
+        doc.set_selection(
+            view.id,
+            Selection::new(selection, doc.selection(view.id).primary_index()),
```

Comment:
If ranges can potentially get merged, can we always use the original selection's primary index safely?

Reply:
Its save here because we just delete the selections exactly (start/end is always just the start/end of the selection range). That means that this particular function can actually never have overlapping deletion ranges (since selections are always mon-overlapping).

Reply:
Ah my bad aplogoues I looked at the wrong code somehow. This is a trick. Here the selection actually is set manually. This works because we call `set_selection` *before* the transaction is applied and if you look at the code we unconditionally call push so the number of ranges doesn't change and the primary index remains valid. 

We only update the selection this way (with head += 1) so that the semantics for `delete_X_forward` work as intended. Just the normally mapping the selection trough the transaction. When you call set_selection I'd normalized si if the selections overlap (because of the head += 1) that's automatically taken care of (and the primary index is then adjusted automatically). 

Furthermore when we apply the transaction below the new selection is still mapped trough the change set so only then becomes then do "overlapping deletions" become a problem. However as long as the selection is valid here (which it is) then that mapping will work too.

Reply:
Ahhh, I get it now, so all the overlapping selections are still actually there at the end of the function, and they get removed during the transaction application. Thanks for bearing with me.